### PR TITLE
Support extracting using companion object apply method.

### PR DESCRIPTION
**Pull request by LATaylor-guardian**
If a class has no constructors, use the companion objects apply method.
To achieve this we undertook some small refactorings.
See tests for details.
This PR allows for deserialization directly to Scrooge generated code as Scrooge does not generate case classes.

### CODE REVIEW

**Comment by LATaylor-guardian** [comment]
@cb372

**Comment by cb372** [comment]
Aw crap, it turns out that `java.lang.reflect.Executable` is only available in Java 8. Should be easy to fix.

File: core/src/main/scala/org/json4s/reflect/Reflector.scala [hunk 1]
```diff
@@ -186,3 +209,59 @@ object Reflector {
     }
 
 }
+
+/**
+ * This class is intended as a workaround until we are able to use Java 8's java.lang.reflect.Executable class.
+ */
+class Executable private (val method: Method, val constructor: Constructor[_]) {
```
Comment (seratch) [inline core/src/main/scala/org/json4s/reflect/Reflector.scala]: How about newly creating `Executable.scala`?

**Comment by seratch** [comment]
Others look good to me. If there is no objection, I'd love to merge this PR.

**Comment by LATaylor-guardian** [comment]
@seratch Thanks, I've updated.

**Comment by LATaylor-guardian** [comment]
Brilliant. @seratch Any chance of getting a new version of json4s released so we can use this straight away?

**Comment by seratch** [comment]
In a few days, I'm going to release the next RC version with these small fixes listed below. https://github.com/json4s/json4s/issues?q=is%3Aopen+milestone%3A3.3+is%3Aissue

If you're in a hurry, please build and publish the latest version into your in-house repository for now.

**Pull request merged**